src/vsr/replica.zig

```diff
@@ -3719,6 +3719,57 @@ pub fn ReplicaType(
+        /// The goal is to repair as much as possible without triggering unnecessary state transfer.
+        ///
+        /// - Repairing uncommitted ops is necessary as primary — we will need them to proceed.
+        /// - Repairing committed + not-yet-checkpointed ops is useful because we might crash,
+        ///   in which case we will need them to recover.
+        /// - Repairing committed + checkpointed ops:
+        ///   - backups don't repair checkpointed ops, since there is no guarantee that they will
+        ///     ever be available (if our head is behind), and we don't want to stall the new view
+        ///     startup.
+        ///   - primaries do repair checkpointed ops — as many as are guaranteed to exist anywhere in
+        ///     the cluster, so that they can help lagging backups catch up.
+        fn op_repair_min(self: *const Self) u64 {
+            assert(self.status == .normal or self.status == .view_change);
+            assert(self.op >= self.op_checkpoint());
+
+            const op = op: {
+                if (self.primary_index(self.view) == self.replica) {
+                    assert(self.status == .normal or self.do_view_change_quorum);
+                    // This is the oldest op that is guaranteed to be in the WALs of any replica.
+                    // (Assuming that this primary has not been superseded.)
+                    break :op std.math.min(
+                        // Add the oldest pipeline_prepare_queue_max ops because they may have been
+                        // newer ops which were then truncated by a view-change, causing the head op
+                        // to backtrack.
+                        self.op + constants.pipeline_prepare_queue_max,
+                        // ...But the pipeline messages could not have moved past the checkpoint.
+                        self.op_checkpoint_trigger(),
+                    ) -| (constants.journal_slot_count - 1);
+                } else {
+                    // Strictly speaking a backup only needs to repair commit_min+1… to proceed.
+                    // However, if the backup crashes and recovers, it will need to replay ops
+                    // from the checkpoint, so we repair slightly more.
+                    if (self.op_checkpoint() == self.op) {
+                        // Don't allow "op_repair_min > op_head".
+                        break :op self.op_checkpoint();
+                    } else {
+                        // +1 because the primary (and other replicas) may overwrite op_checkpoint.
+                        // And we don't need op_checkpoint's prepare/header in our log if we crash.
+                        break :op self.op_checkpoint() + 1;
+                    }
+                }
+            };
+
+            assert(op <= self.op);
+            assert(op <= self.commit_min + 1);
+            assert(op <= self.op_checkpoint() + 1);
+            assert(op >= self.op -| (constants.journal_slot_count - 1));
```

Comment:
micro nit, but maybe phrase this in terms of `self.op - op` -- the lenght of the rage of ops we are repairing?